sunpy/net/dataretriever/sources/gong.py

```diff
@@ -0,0 +1,282 @@
+                    if (6562 <= wave <= 6563):
+                        patterns.append(url_pattern_2)
+                    elif wave == 6768:
+                        patterns.append(url_pattern_1)
+
+
+        #All valid patterns to be downloaded are in the patterns list.
+        instruments_to = list() #The instruments from which user wants to download
+        if not instrument_in:
+            instruments_to.extend(table_instruments)
+        else:
+            instruments_to.append(kwargs['instrument'])
+
+        
+        def download_from_nso(id, ObsID, time_range):
+            #Cannot scrape urls from NSO. The logic for downloading is
+            #dependent on NSO's policy of uploading data. Every day staring from 00:04
+            #fits files are uploaded regularly at an interval of 10 minutes.
+            result = list()
+            base_url = 'ftp://gong2.nso.edu/QR/{id}qa/{date:%Y%m}/{ObsID}{id}qa{date:%y%m%d}/{ObsID}{id}qa{date:%y%m%d}t{date:%H%M}.fits.gz'
+            total_days = (time_range.end - time_range.start).days + 1
+            all_dates = time_range.split(total_days)
+            for day in all_dates:
+                today = datetime.datetime(day.end.year, day.end.month, day.end.day,
+                                          0, 4, 0)
+                tomorrow = today
+                tomorrow += datetime.timedelta(days=1)
+                while (today < tomorrow):
+                    if (time_range.start <= today <= time_range.end):
+                        result.append(base_url.format(id=id, ObsID=ObsID, date=today))
+                    today += datetime.timedelta(seconds = 600)#10 minutes, 600 seconds.
+            result = list(set(result)) #Remove duplicates, for safety.
+            return result
+                
+            
+        
+        for pattern_ in patterns:
+            urls = list()
+            if (pattern_ == url_pattern_1):
+                if not physobs_in:
+                    for instr in instruments_to:
+                        arr = download_from_nso('i',instr,timerange)
+                        urls.extend(arr)
+                        arr = download_from_nso('b',instr,timerange)
+                        urls.extend(arr)
+                else:
+                    for instr in instruments_to:
+                        urls.extend(download_from_nso(table_physobs[kwargs['physobs']],instr,timerange))
+                result.extend(urls)
+            elif (pattern_ == url_pattern_2):
+                urls = [url_pattern_2.format(ObsID=id[0].upper()) for id in instruments_to]
+                arr = [Scraper(pattern__).filelist(timerange) for pattern__ in urls]
+                [result.extend(url) for url in arr if len(url)>0]
+        
+        if not timerange:
+            return []
+        final_result = list()
+        #Check for illegitimate urls.
+        for urls in result:
+            try:
+                check = urllib2.urlopen(urls)
```

Comment:
The ```download_from_nso``` function is only for ftp specifically they are for [site-1](ftp://gong2.nso.edu/QR/bqa/yyyymm/{ObsID}bqayymmdd/) and [site-2](ftp://gong2.nso.edu/QR/iqa/yyyymm/{ObsID}iqayymmdd/). It is tested only on the ftp one, Scraper works fine for the H-alpha site in the spreadsheet. If you're asking, that ```urllib2``` would work on ftp, then yes it would. It can open urls with ftp as well. Could you please elaborate more about your question.